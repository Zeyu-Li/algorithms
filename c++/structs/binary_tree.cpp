/*
* Binary tree implementation 
* Resource: https://gist.github.com/toboqus/def6a6915e4abd66e922
*/
#include <iostream>


// each node struct
// can change it's value(s) below such that 
// the node can encapsulate the data set
struct node {
    int value;

    node *lhs;
    node *rhs;
};

// binary tree class
class Binary_tree {
    public:
        // inits binary tree
        Binary_tree();
        ~Binary_tree();

        // prints (from root)
        // pre-order = Root-Left-Right
        void pre_order();
        // in order = Left-Root-Right
        void in_order();
        // post order = Left-Right-Root
        void post_order();

        // general insert of value (int)
        void insert(int value);

        // searches and returns the node
        node *search(int value);
    private:
        // root node
        node *root;

        // print helper funcs (overloads)
        void pre_order(node *leaf);
        void in_order(node *leaf);
        void post_order(node *leaf);

        // destroyer helper func
        void delete_node(node *some_node);

        // overloading insert of adding root to somewhere in the tree
        void insert(int value, node *leaf);
        
        // searches helper func
        node *search(int value, node *leaf);
};

// constructor and destructor
Binary_tree::Binary_tree() {
    root = nullptr;
}
Binary_tree::~Binary_tree() {
    delete_node(root);
}

// inserts
void Binary_tree::insert(int value) {
    if (root != nullptr) {
        // insert a key at root
        insert(value, root);
    } else {
        // inserting a new node
        root = new node;
        root->lhs = nullptr;
        root->rhs = nullptr;
        root->value = value;
    }
}
void Binary_tree::insert(int value, node *leaf) {
    if (value < leaf->value) {
        // goto left subtree

        // if non trivial, keep recursively finding the subtree
        if (leaf->lhs != nullptr) {
            insert(value, leaf->lhs);
        } else {
            // inserting a new node
            leaf->lhs = new node;
            leaf->lhs->lhs = nullptr;
            leaf->lhs->rhs = nullptr;
            leaf->lhs->value = value;
        }
    } else {
        // goto right subtree

        // if non trivial, keep recursively finding the subtree
        if (leaf->rhs != nullptr) {
            insert(value, leaf->rhs);
        } else {
            // inserting a new node
            leaf->rhs = new node;
            leaf->rhs->lhs = nullptr;
            leaf->rhs->rhs = nullptr;
            leaf->rhs->value = value;
        }
    }
}

// deletes subtrees
void Binary_tree::delete_node(node *some_node) {
    if (some_node != nullptr) {
        // recursive delete on both rhs and lhs
        delete_node(some_node->lhs);
        delete_node(some_node->rhs);
        // delete self root
        delete some_node;
    }
}

// prints
void Binary_tree::pre_order() {
	pre_order(root);
	std::cout << std::endl;
}
void Binary_tree::pre_order(node *leaf) {
	if (leaf != nullptr) {
        std::cout << leaf->value << " ";
        pre_order(leaf->lhs);
        pre_order(leaf->rhs);
    }
}
void Binary_tree::in_order() {
	in_order(root);
	std::cout << std::endl;
}
void Binary_tree::in_order(node *leaf) {
	if (leaf != nullptr) {
        // print lhs then rhs
        in_order(leaf->lhs);
        std::cout << leaf->value << " ";
        in_order(leaf->rhs);
    }
}
void Binary_tree::post_order() {
	post_order(root);
	std::cout << std::endl;
}
void Binary_tree::post_order(node *leaf) {
	if (leaf != nullptr) {
        post_order(leaf->lhs);
        post_order(leaf->rhs);
        std::cout << leaf->value << " ";
    }
}

// searches
node *Binary_tree::search(int value) {
    // search in root
    return search(value, root);
}
node *Binary_tree::search(int value, node *leaf) {
    // if not found, returns null ptr

    // search in root
    if (leaf == nullptr) {
        return nullptr;
    } else {
        if (leaf->value == value) {
            return leaf;
        } else if (leaf->value < value) {
            // goto right sub tree
            return search(value, leaf->rhs);
        } else {
            return search(value, leaf->lhs);
        }
        
    }
    
    return search(value, root);
}
