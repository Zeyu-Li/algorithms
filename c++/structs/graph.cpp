// regular graph representation as a array of pairs (source and desinations) and the weight
// https://gist.github.com/MagallanesFito/791f736a0d21708794aafa11a0416201
#include <iostream>
#include <vector>
#include <utility>

class Graph {
    private:
        int V, E;

        // pointer to vector containing adjacency list
        std::vector<std::pair<int, std::pair<int, int>>> edges;
    public:
        // contructor
        Graph();
        Graph(int V, int E);
        // destructor not needed
        // ~Graph();

        // adds an edge to graph, default weight 1
        void add_edge(int start, int end);
        // overloads with weight
        void add_edge(int start, int end, int weight);

        // print graph to iostream (connection then weight in brackets)
        void print();
};

Graph::Graph(int V, int E) {
    this->V = V;
    this->E = E;
}
Graph::Graph() {
    V = 0;
    E = 0;
}

void Graph::add_edge(int start, int end) {
    edges.push_back({1, {start, end}});
}
void Graph::add_edge(int start, int end, int weight) {
    edges.push_back({weight, {start, end}});
}

void Graph::print(void) {
    std::vector<std::pair<int, std::pair<int, int>>>::iterator it;
    for(it = edges.begin();it!=edges.end();it++){
        std::cout << it->second.first << " - " << it->second.second << " (" << it->first << ")" << std::endl;
    }
}

int main() {
    Graph graph;

    graph.add_edge(2, 4,10);
    graph.add_edge(4,3);
    graph.print();
    return 0;
}

