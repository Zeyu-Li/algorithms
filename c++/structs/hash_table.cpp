/* 
* Hash Table by Andrew Li
*/ 
#include <iostream>
#include <list>


class Hash_table {
    private: 
        // number of rows
        int size;
        // container
        std::list<int> *table;

        // hash function is a simple modulo
        int hashing(int num) {
            return num % size;
        }

    protected:
        // contructor, n is number of rows
        Hash_table(int n);
        // destructor
        ~Hash_table();

        void insert_item(int item);
        // inserting an array of numbers given size and array
        void insert_items(int n, int arr[]);
        // NOTE: only deletes first occurrence of item
        void delete_item(int item);

        // prints
        void print();

        // getter
        int get_size();
};
// contructor
Hash_table::Hash_table(int n) {
    size = n;
    table = new std::list<int>[n];
}
Hash_table::~Hash_table() {
    delete[] table;
}
// insert item via hash func
void Hash_table::insert_item(int item) {
    table[hashing(item)].push_back(item);
}
// overloads with multiple items
void Hash_table::insert_items(int items, int arr[]) {
    for (int i = 0; i<items; i++) {
        table[hashing(arr[i])].push_back(arr[i]);
    }
}
// remove item via hash func
void Hash_table::delete_item(int item) {
    int row = hashing(item);

    // iterate through table by rows
    std::list<int> :: iterator i;
    for (i = table[row].begin(); i != table[row].end(); i++) {
        if (*i == item) break;
    }

    // if it exists in the table, delete it (possible to fit this in the previous for loop)
    if (i != table[row].end()) table[row].erase(i);
}

// print
void Hash_table::print() {
    for (int i = 0; i<size; i++) {
        std::cout << i;
        for (auto item: table[i]) {
            std::cout << " -> " << item;
        }
        std::cout << std::endl;
    }
}

int Hash_table::get_size() {
    return size;
}
