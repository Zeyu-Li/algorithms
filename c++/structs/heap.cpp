/* Heaps */ 
#include <iostream>
#include <algorithm>
#include <vector>

class Min_heap {
    public:
        Min_heap();

        void push_back(int);
        // push back array of ints
        void push_back(int, int []);
        void pop_back();

        // returns value given index
        int get_value(int);
        
        // get size
        int get_size();

        // print
        void print();
    protected:
        int size;
        std::vector<int> heap;
};

Min_heap::Min_heap() {
    size = 0;
}

void Min_heap::push_back(int value) {
    heap.push_back(value);
    if (heap.size() == 1) {
        std::make_heap(heap.begin(), heap.end());
    } else {
        std::push_heap(heap.begin(), heap.end());
    }
}
void Min_heap::push_back(int size, int arr[]) {
    for (int i = 0; i < size; i++) {
        heap.push_back(arr[i]);
    }
    std::make_heap(heap.begin(), heap.end());
}
void Min_heap::pop_back() {
    // if heap is zero, then cannot pop
    if (heap.size() == 0) return;
    // alternately
    // throw 1;
    std::pop_heap(heap.begin(), heap.end());
    heap.pop_back();
}

int Min_heap::get_value(int index) {
    return heap[index];
}
int Min_heap::get_size() {
    return size;
}

// print
void Min_heap::print() {
    for (unsigned i=0; i<heap.size(); i++) std::cout << heap[i] << ' ' ;
    std::cout << std::endl;
}
