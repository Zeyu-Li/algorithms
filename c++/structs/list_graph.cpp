/* 
* Adjacency List Graph
* Modified from: https://www.geeksforgeeks.org/graph-and-its-representations/ & from hash table
*/ 
#include <iostream>
#include <vector>
// remove if necessary
#pragma once


class Graph_l {
    protected:
        bool directed;
        // hash table list
        std::vector<int> *adj;
        int size;
        bool flag = false;
    public:
        // contructor, first arg directed or undirected, next is the number of vertices
        Graph_l(bool, int);
        // overloads main contructor, default undirected, arg is the number of vertices
        Graph_l(int);
        // destructor
        ~Graph_l();

        // add edge
        void add_edge(int source, int destination);
        // TODO: add list of pairs?
        // void add_edges(std::pairs);
        void delete_edge(int source, int destination);

        // gets size of graph (ie number of vertices)
        int get_size();
        // print
        void print();
        // print all edges to a vertix
        void print_edges(int);
};

Graph_l::Graph_l(int v) {
    directed = true;
    size = v;
    adj = new std::vector<int>[v];
}
Graph_l::Graph_l(bool dir, int v) {
    directed = dir;
    size = v;
    adj = new std::vector<int>[v];
}
Graph_l::~Graph_l() {
    delete[] adj;
}

void Graph_l::add_edge(int source, int destination) {
    adj[source].push_back(destination);
    // if undirected, also do the reverse
    if (!directed) adj[destination].push_back(source);
}

void Graph_l::delete_edge(int source, int destination) {
    // iterate through table by rows
    std::vector<int> :: iterator i;
    for (i = adj[source].begin(); i != adj[source].end(); i++) {
        if (*i == destination) break;
    }

    if (i != adj[source].end()) adj[source].erase(i);
    // if undirected, do the reverse
    if (!directed && !flag) {
        flag = !flag;
        delete_edge(destination, source);
        flag = !flag;
    }
}

void Graph_l::print() {
    for (int i = 0; i<size; i++) {
        std::cout << i;
        for (auto item: adj[i]) {
            std::cout << " -> " << item;
        }
        std::cout << std::endl;
    }
}
void Graph_l::print_edges(int edge) {
    std::cout << edge;
    for (auto item: adj[edge]) {
        std::cout << " -> " << item;
    }
    std::cout << std::endl;
}
int Graph_l::get_size() {
    return size;
}
