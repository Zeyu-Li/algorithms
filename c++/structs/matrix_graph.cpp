/* 
* Matrix/Table Graph
* Modified from: https://github.com/DorianDepriester/adjacencyMatrix
*/
#include <iostream>
#include <vector>

class Graph_m {
    public:
        // contructors, default undirected
        Graph_m();
        // contructor with directness boolean 
        Graph_m(bool);
        // contructor with size (number of vertices)
        Graph_m(int);
        Graph_m(bool, int);
        // destructor
        ~Graph_m();

        // print
        void print();
        int get_size();

        // resize
        void resize(int new_size);

        // setter (edge)
        void add_edge(int, int);
        // overloads when default weight is not 1
        void add_edge(int, int, int);

        // main matrix, therefore getter unnecessary
        std::vector<std::vector <int>> matrix;
    protected:
        int size;
        bool directed;
};

Graph_m::Graph_m() {
    size=0;
    directed = true;
}
Graph_m::Graph_m(bool direc) {
    size=0;
    directed = direc;
}
Graph_m::Graph_m(int n) {
    directed = true;
    // resize
    size=0;
    resize(n);
    size=n;
}
Graph_m::Graph_m(bool direc, int n) {
    directed = direc;
    // resize
    size=0;
    resize(n);
    size=n;
}
Graph_m::~Graph_m() {
    size=0;
}

// resize
void Graph_m::resize(int new_size) {
    if (new_size > size) {
        // enlarge the matrix
        for (int i = size; i<new_size; i++) {
            // add zero to each row
            // zeros for each row
            std::vector<int> bottom;
            for (int j = size; j<new_size; j++) {
                bottom.push_back(0);
            }
            matrix.push_back(bottom);
        }
    }
}

// add edge row then column
void Graph_m::add_edge(int y, int x) {
    // sets size as the max of x or y
    int max = x > y ? x : y;
    if (max>size) {
        size = max;
        // Resize
        resize(max);
    }
    matrix[y][x] = 1;
    // if undirected
    if (!directed) matrix[x][y] = 1;
}
void Graph_m::add_edge(int y, int x, int weight) {
    // sets size as the max of x or y
    int max = x > y ? x : y;
    if (max>size) {
        size = max;
    }
    matrix[y][x] = weight;
    // if undirected
    if (!directed) matrix[x][y] = weight;
}

int Graph_m::get_size(void) {
    return size;
}

void Graph_m::print(void) {
    int n = size;
    for (int i=0;i<n;i++) {
        for (int j=0;j<n;j++) {
            std::cout << matrix[i][j]<< "\t";
        }
        std::cout << std::endl;
    }
}
