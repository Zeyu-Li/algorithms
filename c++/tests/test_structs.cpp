/* 
* Testing Structs
* Description: A test for all the classes
*/ 
#include <iostream>
#include "../structs/binary_tree.cpp"
#include "../structs/heap.cpp"
#include "../structs/hash_table.cpp"
#include "../structs/graph.cpp"
#include "../structs/list_graph.cpp"
#include "../structs/matrix_graph.cpp"

using namespace std;

int main() {
    /* binary tree */
    cout << "Binary tree:" << endl;
    Binary_tree *tree = new Binary_tree();

    // for best results, randomize inserts so the tree can be somewhat balanced
    tree->insert(10);
    tree->insert(5);
    tree->insert(11);
    tree->insert(47);
    tree->insert(0);

    tree->pre_order();
    tree->in_order();
    tree->post_order();

    node *find = tree->search(10);
    // must set overload to public to do the thing below
    // tree->post_order(find);

    delete tree;

    /* heap (very rudimentary) */
    cout << "Heap:" << endl;
    int arr[] = {1,5,7,3,68,4};
    Min_heap heap;

    heap.push_back(sizeof(arr)/sizeof(arr[0]), arr);
    heap.push_back(4);
    heap.pop_back();
    heap.print();

    /* hash table */
    cout << "Hash table:" << endl;
    Hash_table table(3);

    int a[] = {2,5,7,3,7,3,5,2,6,3};
    int size = sizeof(a)/sizeof(a[0]);

    table.insert_items(size, a);
    table.delete_item(7);

    // prints
    table.print();

    /* graph */
    cout << "Graph:" << endl;
    Graph graph;

    graph.add_edge(2, 4, 10);
    graph.add_edge(4,3);
    graph.print();

    /* list graph */
    cout << "List graph:" << endl;
    Graph_l graph_l(false, 4);
    graph_l.add_edge(1,3);
    graph_l.add_edge(0,2);
    graph_l.add_edge(0,1);
    graph_l.delete_edge(0,2);
    graph_l.print();
    graph_l.add_edge(0,2);
    graph_l.print_edges(0);

    /* matrix graph */
    cout << "Matrix graph:" << endl;
    Graph_m graph_m(false, 5);
    // std::cout << graph.get_size();
    graph_m.add_edge(3,2);
    graph_m.add_edge(2,1, 5);

    graph_m.print();
    // std::cout << graph.matrix[1][1];

    return 0;
}
