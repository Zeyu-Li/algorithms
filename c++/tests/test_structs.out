Binary tree:
10 5 0 11 47
0 5 10 11 47
0 5 47 11 10
Heap:
7 5 4 3 1 4
Hash table:
0 -> 3 -> 3 -> 6 -> 3
1 -> 7
2 -> 2 -> 5 -> 5 -> 2
Graph:
2 - 4 (10)
4 - 3 (1)
List graph:
0 -> 1
1 -> 3 -> 0
2
3 -> 1
0 -> 1 -> 2
Matrix graph:
0       0       0       0       0
0       0       5       0       0
0       5       0       1       0
0       0       1       0       0
0       0       0       0       0